espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/handlers/MobileBackdoor.java

```diff
@@ -0,0 +1,110 @@
+package io.appium.espressoserver.lib.handlers;
+
+import android.app.Activity;
+import android.util.ArrayMap;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException;
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidArgumentException;
+import io.appium.espressoserver.lib.helpers.InvocationOperation;
+import io.appium.espressoserver.lib.model.MobileBackdoorMethod;
+import io.appium.espressoserver.lib.model.MobileBackdoorParams;
+
+import static io.appium.espressoserver.lib.helpers.AndroidLogger.logger;
+
+public class MobileBackdoor implements RequestHandler<MobileBackdoorParams, String> {
+
+    @Override
+    public String handle(final MobileBackdoorParams params) throws AppiumException {
+        logger.info("Invoking Backdoor");
+        Activity activity = getCurrentActivity();
+
+        List<InvocationOperation> ops = getBackdoorOperations(params);
+
+        if (ops.isEmpty()) {
+            throw new InvalidArgumentException("Please pass name(s) of method to be invoked");
+        }
+
+//        First try to find the method in Application object
+        Object invocationResult = safeInvokeOnApplication(activity, ops);
+
+        // if backdoor method not found in Application, try to find the method in urrent Activity object
+        if (invocationResult == null) {
+            invocationResult = invokeOnActivity(activity, ops);
+        }
+
+        return invocationResult.toString();
+    }
+
+    private Object invokeOnActivity(Activity activity, List<InvocationOperation> ops) throws AppiumException {
+        Object invocationResult = null;
+        Object invokeOn = activity;
+        for (InvocationOperation op : ops) {
+            try {
+                invocationResult = op.apply(invokeOn);
+                invokeOn = invocationResult;
+            } catch (Exception e) {
+                throw new AppiumException(e);
+            }
+        }
+        return invocationResult;
+    }
+
+    private Object safeInvokeOnApplication(Activity activity, List<InvocationOperation> ops) {
```

Comment:
please mark as Nullable

Reply:
I see this method mostly repeats the previous one. Can we refactor them to avoid duplication?